bears/c_languages/codeclone_detection/ClangComplexityBear.py

```diff
@@ -0,0 +1,88 @@
+from clang.cindex import Index, CursorKind
+
+from coalib.bears.LocalBear import LocalBear
+from coalib.results.Result import Result
+from bears.c_languages.ClangBear import clang_available
+
+
+class ClangComplexityBear(LocalBear):
+    """
+    Calculates cyclomatic complexity of each function and displays it to the
+    user.
+    """
+    check_prerequisites = classmethod(clang_available)
+    decisive_cursor_kinds = {
+        CursorKind.IF_STMT, CursorKind.WHILE_STMT, CursorKind.FOR_STMT,
+        CursorKind.DEFAULT_STMT, CursorKind.CASE_STMT}
+
+    def run(self, filename, file, max_complexity: int=10):
+        """
+        Calculates cyclomatic complexity of functions in file.
+
+        :param max_complexity:  Maximum cyclomatic complexity that is
+                                considered to be normal. The value of 10 had
+                                received substantial corroborating evidence.
+                                But the general recommendation: "For each
+                                module, either limit cyclomatic complexity to
+                                [the agreed-upon limit] or provide a written
+                                explanation of why the limit was exceeded."
+        """
+        self.complexities = {}
+        counter = {'decisions': 0, 'exits': 0}
+
+        def _traverse(cursor, filename, top_function_level=False):
+            """
+            Traversing clang AST to calculate cyclomatic complexity.
+
+            :param cursor:             The node to print. (Clang cursor.)
+            :param filename:           Name of the file being processed.
+            :param top_function_level: Whether cursor is in the top level of
+                                       the function.
+            """
+            file = cursor.location.file
+
+            if file is not None and file.name == filename:
+                cursor_kind = cursor.kind
+            else:
+                cursor_kind = None
```

Comment:
Not really. The traversal starts in another file so it's necessary to get to out file. Below we're traversing all the children, anyway.